flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecSortMergeJoin.java

```diff
@@ -0,0 +1,186 @@
+                        .getBytes();
+        int externalBufferNum = 1;
+        if (joinType == FlinkJoinType.FULL) {
+            externalBufferNum = 2;
+        }
+
+        long managedMemory = externalBufferMemory * externalBufferNum + sortMemory * 2;
+
+        SortCodeGenerator leftSortGen = newSortGen(config, leftKeys, lType);
+        SortCodeGenerator rightSortGen = newSortGen(config, rightKeys, rType);
+
+        int[] keyPositions = IntStream.range(0, leftKeys.length).toArray();
+        SortMergeJoinOperator operator =
+                new SortMergeJoinOperator(
+                        1.0 * externalBufferMemory / managedMemory,
+                        joinType,
+                        leftIsSmaller,
+                        condFunc,
+                        ProjectionCodeGenerator.generateProjection(
+                                new CodeGeneratorContext(config),
+                                "SMJProjection",
+                                lType,
+                                keyType,
+                                leftKeys),
+                        ProjectionCodeGenerator.generateProjection(
+                                new CodeGeneratorContext(config),
+                                "SMJProjection",
+                                rType,
+                                keyType,
+                                rightKeys),
+                        leftSortGen.generateNormalizedKeyComputer("LeftComputer"),
+                        leftSortGen.generateRecordComparator("LeftComparator"),
+                        rightSortGen.generateNormalizedKeyComputer("RightComputer"),
+                        rightSortGen.generateRecordComparator("RightComparator"),
+                        newSortGen(config, keyPositions, keyType)
+                                .generateRecordComparator("KeyComparator"),
+                        filterNulls);
+
+        TwoInputTransformation<RowData, RowData, RowData> ret =
+                ExecNodeUtil.createTwoInputTransformation(
+                        lInput,
+                        rInput,
+                        getDesc(),
+                        SimpleOperatorFactory.of(operator),
+                        InternalTypeInfo.of(getOutputType()),
+                        rInput.getParallelism(),
+                        managedMemory);
+
+        if (inputsContainSingleton()) {
+            ret.setParallelism(1);
+            ret.setMaxParallelism(1);
+        }
+        return ret;
+    }
+
+    private SortCodeGenerator newSortGen(TableConfig config, int[] originalKeys, RowType type) {
+        boolean[] originalOrders = new boolean[originalKeys.length];
+        Arrays.fill(originalOrders, true);
+        boolean[] nullsIsLast = SortUtil.getNullDefaultOrders(originalOrders);
+        Tuple3<int[], boolean[], boolean[]> ret =
+                SortUtil.deduplicateSortKeys(originalKeys, originalOrders, nullsIsLast);
```

Comment:
add some comments to explain what's each element in Tuple3?

Reply:
return `SortSpec` directly ? even we can use `SortSpec` in `SortCodeGenerator` ?